### DISCUSSION TITLE
Is it intentional that facade and helper function differ in triggering beforeResolving event?

### DISCUSSION THREAD

**Discussion by nanaya** [laravel/framework#49255] [Q&A] [answered]
Given this in app service provider's register:

```
$this->app->beforeResolving(function () {
  echo 'before resolving something'.PHP_EOL;
});
```

I'm getting different results between using `app('session')`/`session()` helpers and `Session` facade:

```
> Session::get('x')
before resolving something
before resolving something
before resolving something
= null

> app('session')->get('x')
before resolving something
= null

> session()->get('x')
before resolving something
= null

> Session::get('x')
= null

> session()->get('x')
before resolving something
= null
```

Is this intentional that facade only triggers the event on first usage? (or that `app` resolver triggers the event every time for singleton)

**Comment by crynobone** [comment]
All facades behaves like a singleton and will only resolved once during the request.

**Reply by nanaya** [reply]
but when accessed through helper function the event triggers every time?

**Reply by crynobone** [reply]
because that doesn't go though facades

**Reply by nanaya** [reply]
but the [documentation](https://laravel.com/docs/10.x/facades#facades-vs-helper-functions) says:

> There is absolutely no practical difference between facades and helper functions.

so is this incorrect?

**Reply by crynobone** [reply]
The line still makes sense to me since the context here is during testing where both would almost identical.

**Reply by nanaya** [reply]
so it's intentional that using `app()` or helper functions (`session()`, `auth()`, etc) triggers beforeResolving event while facade (`Session`, `Auth`) doesn't?

**Reply by crynobone** [reply]
That's not correct. Helpers and Facades will get resolved through the IoC, If the binding is Singleton, then it will resolved once per request otherwise it will resolved each time you call it. Facade at the same time, cached the resolved instance and that make it behave as a singleton. Your example above prove this behavior

**Reply by nanaya** [reply]
I've been talking in context of second and later calls to the corresponding functions/facades in same request...

In term of code, this:
```
session()->get('x');
session()->get('x');
```
Triggers `beforeResolving` for 'session' twice.

Whereas this:
```
Session::get('x');
Session::get('x');
```
Only triggers `beforeResolving` for 'session' once.

So this is correct and intentional, right?

**Reply by crynobone** [reply, marked as answer]
Yes, that's correct.
